# Applied Data Science Capstone Project
Salvador Roca Guillén

## 1. Importing libraries

First of all, we will import (and install, if needed) the libraries needed for this assignment.

In [1]:
import numpy as np # library to handle data in a vectorized manner

import pandas as pd # library for data analsysis
pd.set_option('display.max_columns', None)
pd.set_option('display.max_rows', None)

from geopy.geocoders import Nominatim # convert an address into latitude and longitude values

import requests # library to handle requests

# Matplotlib and associated plotting modules
import matplotlib.cm as cm
import matplotlib.colors as colors

# import k-means from clustering stage
from sklearn.cluster import KMeans

import csv # library to handle CSV

!pip install beautifulsoup4 # library for HTML data scraping

import folium # map rendering library

print('Libraries imported.')

Libraries imported.


## 2. Download and Explore Dataset

After importing the needed libraries, we need to obtain the dataset by scraping the data from the next link: https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M

To do that, we will use BeautifulSoup 4 library, which is the one recommended in the assignment, and has been imported in the previous cells of this notebook.

In [2]:
from bs4 import BeautifulSoup
from urllib.request import urlopen


# Fetch the html file
html_doc = urlopen('https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M').read()

# Parse the html file
soup = BeautifulSoup(html_doc, 'html.parser')

# Format the parsed html file
strhtm = soup.prettify()

#  Find table in HTML code
table = soup.find('table')
headers = [th.text.encode("utf-8") for th in table.select("tr th")]
with open("out.csv", "w") as f:
    wr = csv.writer(f)
    wr.writerow(headers)
    wr.writerows([[td.text.encode("utf-8") for td in row.find_all("td")] for row in table.select("tr + tr")])

df = pd.read_csv("out.csv")

After executing this, we will find a new .csv file with the raw information needed for the assignment.

## 3. Dataset pre-processing

It can be seen in the .csv file that there are some special characters that need to be cleaned. In addition, the assignment requires:
- Ignore the rows with "Not assigned" Borough
- Assign the Borough name to the Neighborhoods with "Not assigned" value
- Group all neighborhoods by Postal Code
- Print dataframe shape, in order to know number of rows and columns it contains

In [3]:
# Clean special characters out of dataframe
df.columns=['Postal Code', 'Borough', 'Neighborhood']
df = df.replace('b\'','', regex=True)
df = df.replace('b"','', regex=True)
df = df.replace('"','', regex=True)
df = df.replace('\'', '', regex=True)
df = df.replace(r'\\n','', regex=True)

# Remove all rows with "Not assigned" Borough
df = df[df.Borough != 'Not assigned']

# Assign Borough name to "Not assigned" Neighborhood
df.loc[df['Neighborhood']=='Not assigned', 'Neighborhood'] = df['Borough']

# Group all neighborhoods in the same Postal Code
df = df.groupby(['Postal Code', 'Borough'], as_index=False).agg(', '.join)

# Print the shape of the dataframe
df.shape

(103, 3)

With this, we now know the dataframe consists of 103 rows and 3 columns. The entire dataframe is printed below to verify all the information is consistent.

In [4]:
# Print the entire dataframe to check if data has been correctly processed
df

Postal Code           Borough  \
0           M1B       Scarborough   
1           M1C       Scarborough   
2           M1E       Scarborough   
3           M1G       Scarborough   
4           M1H       Scarborough   
5           M1J       Scarborough   
6           M1K       Scarborough   
7           M1L       Scarborough   
8           M1M       Scarborough   
9           M1N       Scarborough   
10          M1P       Scarborough   
11          M1R       Scarborough   
12          M1S       Scarborough   
13          M1T       Scarborough   
14          M1V       Scarborough   
15          M1W       Scarborough   
16          M1X       Scarborough   
17          M2H        North York   
18          M2J        North York   
19          M2K        North York   
20          M2L        North York   
21          M2M        North York   
22          M2N        North York   
23          M2P        North York   
24          M2R        North York   
25          M3A        North York   
26          M3B        North York   
27          M3C        North York   
28          M3H        North York   
29          M3J        North York   
30          M3K        North York   
31          M3L        North York   
32          M3M        North York   
33          M3N        North York   
34          M4A        North York   
35          M4B         East York   
36          M4C         East York   
37          M4E      East Toronto   
38          M4G         East York   
39          M4H         East York   
40          M4J         East York   
41          M4K      East Toronto   
42          M4L      East Toronto   
43          M4M      East Toronto   
44          M4N   Central Toronto   
45          M4P   Central Toronto   
46          M4R   Central Toronto   
47          M4S   Central Toronto   
48          M4T   Central Toronto   
49          M4V   Central Toronto   
50          M4W  Downtown Toronto   
51          M4X  Downtown Toronto   
52          M4Y  Downtown Toronto   
53          M5A  Downtown Toronto   
54          M5B  Downtown Toronto   
55          M5C  Downtown Toronto   
56          M5E  Downtown Toronto   
57          M5G  Downtown Toronto   
58          M5H  Downtown Toronto   
59          M5J  Downtown Toronto   
60          M5K  Downtown Toronto   
61          M5L  Downtown Toronto   
62          M5M        North York   
63          M5N   Central Toronto   
64          M5P   Central Toronto   
65          M5R   Central Toronto   
66          M5S  Downtown Toronto   
67          M5T  Downtown Toronto   
68          M5V  Downtown Toronto   
69          M5W  Downtown Toronto   
70          M5X  Downtown Toronto   
71          M6A        North York   
72          M6B        North York   
73          M6C              York   
74          M6E              York   
75          M6G  Downtown Toronto   
76          M6H      West Toronto   
77          M6J      West Toronto   
78          M6K      West Toronto   
79          M6L        North York   
80          M6M              York   
81          M6N              York   
82          M6P      West Toronto   
83          M6R      West Toronto   
84          M6S      West Toronto   
85          M7A       Queens Park   
86          M7R       Mississauga   
87          M7Y      East Toronto   
88          M8V         Etobicoke   
89          M8W         Etobicoke   
90          M8X         Etobicoke   
91          M8Y         Etobicoke   
92          M8Z         Etobicoke   
93          M9A         Etobicoke   
94          M9B         Etobicoke   
95          M9C         Etobicoke   
96          M9L        North York   
97          M9M        North York   
98          M9N              York   
99          M9P         Etobicoke   
100         M9R         Etobicoke   
101         M9V         Etobicoke   
102         M9W         Etobicoke   

                                          Neighborhood  
0                                       Rouge, Malvern  
1               Highland Creek, Rouge Hil

## 4. Adding geospatial data and creating Toronto map

To assign Latitude and Longitude to each Postal Code, we will download the .csv file provided in the assignment.

In [5]:
!wget -q -O 'geospatial_data.csv' http://cocl.us/Geospatial_data
df_geo = pd.read_csv("geospatial_data.csv")

Once we have a dataframe containing the information in the .csv, we will merge both dataframes directly, as all the information is sorted in alphabetical ascending order by Postal Code.

In [6]:
df = pd.merge(df, df_geo)
df.head()

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Rouge, Malvern  43.806686   
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [7]:
# Get Toronto latitude and longitude to set center of the map
address = 'Toronto, Canada'

geolocator = Nominatim()
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude

# Create map of Toronto using latitude and longitude values
map_toronto = folium.Map(location=[latitude, longitude], zoom_start=11)

# Add markers to map
for lat, lng, postalcode, borough, neighborhood in zip(df['Latitude'], df['Longitude'], df['Postal Code'], df['Borough'], df['Neighborhood']):
    label = '{} - {}, {}'.format(neighborhood, borough, postalcode)
    label = folium.Popup(label, parse_html=True)
    folium.CircleMarker(
        [lat, lng],
        radius=5,
        popup=label,
        color='blue',
        fill=True,
        fill_color='#3186cc',
        fill_opacity=0.7,
        parse_html=False).add_to(map_toronto)  

map_toronto

## 5. Exploring and analyzing the neighborhoods with Foursquare data

We will use the next function to analyze all the neighborhoods in Toronto area.

In [8]:
CLIENT_ID = 'OYTC5ITWBG1R4XIWJKQ2QQYPY0BOQKTNOEO3HDYTPGIQJJ3Q' # your Foursquare ID
CLIENT_SECRET = 'JZFSM4WGITFCRASKFBKPOSJFA55RPIHDUHDVIRZ4FEDAXXO4' # your Foursquare Secret
VERSION = '20180605' # Foursquare API version

LIMIT = 100

def getNearbyVenues(names, latitudes, longitudes, radius=500):
    
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])

    nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

toronto_venues = getNearbyVenues(names=df['Neighborhood'],
                                   latitudes=df['Latitude'],
                                   longitudes=df['Longitude']
                                  )

Rouge, Malvern
Highland Creek, Rouge Hill, Port Union
Guildwood, Morningside, West Hill
Woburn
Cedarbrae
Scarborough Village
East Birchmount Park, Ionview, Kennedy Park
Clairlea, Golden Mile, Oakridge
Cliffcrest, Cliffside, Scarborough Village West
Birch Cliff, Cliffside West
Dorset Park, Scarborough Town Centre, Wexford Heights
Maryvale, Wexford
Agincourt
Clarks Corners, Sullivan, Tam OShanter
Agincourt North, LAmoreaux East, Milliken, Steeles East
LAmoreaux West
Upper Rouge
Hillcrest Village
Fairview, Henry Farm, Oriole
Bayview Village
Silver Hills, York Mills
Newtonbrook, Willowdale
Willowdale South
York Mills West
Willowdale West
Parkwoods
Don Mills North
Flemingdon Park, Don Mills South
Bathurst Manor, Downsview North, Wilson Heights
Northwood Park, York University
CFB Toronto, Downsview East
Downsview West
Downsview Central
Downsview Northwest
Victoria Village
Woodbine Gardens, Parkview Hill
Woodbine Heights
The Beaches
Leaside
Thorncliffe Park
East Toronto
The Danforth West, Riv

#### Let's find out how many unique categories can be curated from all the returned venues

In [9]:
print('There are {} uniques categories.'.format(len(toronto_venues['Venue Category'].unique())))

There are 284 uniques categories.


#### Analysis of every neighborhood with one-hot encoding

In [10]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")

# add neighborhood column back to dataframe
toronto_onehot['Neighborhood'] = toronto_venues['Neighborhood'] 

# move neighborhood column to the first column
toronto_onehot = toronto_onehot.set_index('Neighborhood')
toronto_onehot = toronto_onehot.reset_index()

toronto_onehot.head()

Neighborhood  Accessories Store  Adult Boutique  \
0                          Rouge, Malvern                  0               0   
1  Highland Creek, Rouge Hill, Port Union                  0               0   
2  Highland Creek, Rouge Hill, Port Union                  0               0   
3  Highland Creek, Rouge Hill, Port Union                  0               0   
4       Guildwood, Morningside, West Hill                  0               0   

   Afghan Restaurant  Airport  Airport Food Court  Airport Gate  \
0                  0        0                   0             0   
1                  0        0                   0             0   
2                  0        0                   0             0   
3                  0        0                   0             0   
4                  0        0                   0             0   

   Airport Lounge  Airport Service  Airport Terminal  American Restaurant  \
0               0                0                 0                    0   
1               0                0                 0                    0   
2               0                0                 0                    0   
3               0                0                 0                    0   
4               0                0                 0                    0   

   Antique Shop  Aquarium  Argentinian Restaurant  Art Gallery  Art Museum  \
0             0         0                       0            0           0   
1             0         0                       0            0           0   
2             0         0                       0            0           0   
3             0         0                       0            0           0   
4             0         0                       0            0           0   

   Arts & Crafts Store  Asian Restaurant  Athletics & Sports  Auto Garage  \
0                    0                 0                   0            0   
1                    0                 0                   0            0   
2                    0                 0                   0            0   
3                    0                 0                   0            0   
4                    0                 0                   0            0   

   Auto Workshop  BBQ Joint  Baby Store  Bagel Shop  Bakery  Bank  Bar  \
0              0          0           0           0       0     0    0   
1              0          0           0           0       0     0    1   
2              0          0           0           0       0     0    0   
3              0          0           0           0       0     0    0   
4              0          0           0           0       0     0    0   

   Baseball Field  Baseball Stadium  Basketball Court  Basketball Stadium  \
0               0                 0                 0                   0   
1               0                 0                 0                   0   
2               0                 0                 0                   0   
3               0                 0                 0                   0   
4               0                 0                 0                   0   

   Beach  Beer Bar  Beer Store  Belgian Restaurant  Bike Shop  Bistro  \
0      0         0           0                   0          0       0   
1      0         0           0                   0          0       0   
2      0         0           0                   0          0       0   
3      0         0           0                   0          0       0   
4      0         0           0                   0          0       0   

   Boat or Ferry  Bookstore  Boutique  Brazilian Restaurant  Breakfast Spot  \
0              0          0         0                     0               0   
1              0          0         0                     0               0   
2              0          0         0                     0               0   
3              0          0         0                     0               0   
4       

And let's examine the new dataframe size.

In [11]:
toronto_onehot.shape

(2248, 284)

#### Next, let's group rows by neighborhood and by taking the mean of the frequency of occurrence of each category

In [12]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped

Neighborhood  Accessories Store  \
0                            Adelaide, King, Richmond           0.010000   
1                                           Agincourt           0.000000   
2   Agincourt North, LAmoreaux East, Milliken, Ste...           0.000000   
3   Albion Gardens, Beaumond Heights, Humbergate, ...           0.000000   
4                              Alderwood, Long Branch           0.000000   
5     Bathurst Manor, Downsview North, Wilson Heights           0.000000   
6                                     Bayview Village           0.000000   
7                   Bedford Park, Lawrence Manor East           0.000000   
8                                         Berczy Park           0.000000   
9                         Birch Cliff, Cliffside West           0.000000   
10  Bloordale Gardens, Eringate, Markland Wood, Ol...           0.000000   
11       Brockton, Exhibition Place, Parkdale Village           0.000000   
12  Business Reply Mail Processing Centre 969 Eastern           0.000000   
13                        CFB Toronto, Downsview East           0.000000   
14  CN Tower, Bathurst Quay, Island airport, Harbo...           0.000000   
15                        Cabbagetown, St. James Town           0.000000   
16                                Caledonia-Fairbanks           0.000000   
17              Canada Post Gateway Processing Centre           0.000000   
18                                          Cedarbrae           0.000000   
19                                 Central Bay Street           0.000000   
20          Chinatown, Grange Park, Kensington Market           0.000000   
21                                           Christie           0.000000   
22                               Church and Wellesley           0.000000   
23                    Clairlea, Golden Mile, Oakridge           0.000000   
24             Clarks Corners, Sullivan, Tam OShanter           0.000000   
25    Cliffcrest, Cliffside, Scarborough Village West           0.000000   
26  Cloverdale, Islington, Martin Grove, Princess ...           0.000000   
27                     Commerce Court, Victoria Hotel           0.000000   
28                                         Davisville           0.000000   
29                                   Davisville North           0.000000   
30  Deer Park, Forest Hill SE, Rathnelly, South Hi...           0.000000   
31     Del Ray, Keelesdale, Mount Dennis, Silverthorn           0.000000   
32           Design Exchange, Toronto Dominion Centre           0.000000   
33                                    Don Mills North           0.000000   
34  Dorset Park, Scarborough Town Centre, Wexford ...           0.000000   
35                       Dovercourt Village, Dufferin           0.000000   
36                                  Downsview Central           0.000000   
37                                Downsview Northwest           0.000000   
38                                     Downsview West           0.000000   
39                 Downsview, North Park, Upwood Park           0.000000   
40        East Birchmount Park, Ionview, Kennedy Park           0.000000   
41                                       East Toronto           0.000000   
42                                   Emery, Humberlea           0.000000   
43                       Fairview, Henry Farm, Oriole           0.000000   
44             First Canadian Place, Underground city           0.000000   
45                   Flemingdon Park, Don Mills South           0.000000   
46                Forest Hill North, Forest Hill West           0.000000   
47                                          Glencairn           0.000000   
48                  Guildwood, Morningside, West Hill           0.000000   
49                     Harbord, University of Toronto           0.000000   
50  Harbourfront East, Toronto Islands, Union Station           0.000000   
51                          Harbourfront, Regent Park           0.000000   
52           

#### Let's represent the top 10 most common venues in each neighborhood into a *pandas* dataframe

In [13]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                           Adelaide, King, Richmond           Coffee Shop   
1                                          Agincourt                Lounge   
2  Agincourt North, LAmoreaux East, Milliken, Ste...            Playground   
3  Albion Gardens, Beaumond Heights, Humbergate, ...         Grocery Store   
4                             Alderwood, Long Branch           Pizza Place   

  2nd Most Common Venue 3rd Most Common Venue        4th Most Common Venue  \
0            Steakhouse                  Café              Thai Restaurant   
1        Breakfast Spot          Skating Rink               Clothing Store   
2                  Park           Yoga Studio  Eastern European Restaurant   
3           Pizza Place   Fried Chicken Joint                  Coffee Shop   
4              Pharmacy                   Gym               Sandwich Place   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Restaurant                   Gym          Burger Joint   
1     Electronics Store      Doner Restaurant            Donut Shop   
2        Discount Store               Dog Run      Doner Restaurant   
3        Sandwich Place  Fast Food Restaurant            Beer Store   
4           Coffee Shop          Skating Rink                   Pub   

  8th Most Common Venue 9th Most Common Venue       10th Most Common Venue  
0                   Bar   American Restaurant                       Bakery  
1             Drugstore   Dumpling Restaurant  Eastern European Restaurant  
2            Donut Shop             Drugstore          Dumpling Restaurant  
3              Pharmacy          Gourmet Shop                  Event Space  
4                  Pool      Doner Restaurant           Dim Sum Restaurant

## 5. Clustering the neighborhoods

Run *k*-means to cluster the neighborhood into 5 clusters.

In [14]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_

array([0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0,
       0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0,
       0, 2, 0, 0, 0, 0, 1, 0, 1, 4, 3, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 0,
       1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1], dtype=int32)

Let's create a new dataframe that includes the cluster as well as the top 10 venues for each neighborhood.

In [15]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = df

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

# delete rows with no cluster label assigned as no data was found in Foursquare database
toronto_merged = toronto_merged[np.isfinite(toronto_merged['Cluster Labels'])]
toronto_merged = toronto_merged.reset_index(drop=True)
toronto_merged['Cluster Labels'] = toronto_merged['Cluster Labels'].astype(int)

toronto_merged.head() # check the last columns!

Postal Code      Borough                            Neighborhood   Latitude  \
0         M1B  Scarborough                          Rouge, Malvern  43.806686   
1         M1C  Scarborough  Highland Creek, Rouge Hill, Port Union  43.784535   
2         M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3         M1G  Scarborough                                  Woburn  43.770992   
4         M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  Cluster Labels 1st Most Common Venue 2nd Most Common Venue  \
0 -79.194353               3  Fast Food Restaurant           Yoga Studio   
1 -79.160497               0        History Museum         Moving Target   
2 -79.188711               0        Medical Center          Intersection   
3 -79.216917               0           Coffee Shop     Korean Restaurant   
4 -79.239476               0    Athletics & Sports      Hakka Restaurant   

  3rd Most Common Venue 4th Most Common Venue        5th Most Common Venue  \
0     Electronics Store               Dog Run             Doner Restaurant   
1                   Bar           Yoga Studio  Eastern European Restaurant   
2           Pizza Place                   Spa               Breakfast Spot   
3      Insurance Office     Electronics Store                      Dog Run   
4  Caribbean Restaurant       Thai Restaurant                       Bakery   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0            Donut Shop             Drugstore   Dumpling Restaurant   
1               Dog Run      Doner Restaurant            Donut Shop   
2   Rental Car Location     Electronics Store    Mexican Restaurant   
3      Doner Restaurant            Donut Shop             Drugstore   
4                  Bank   Fried Chicken Joint   Dumpling Restaurant   

         9th Most Common Venue       10th Most Common Venue  
0  Eastern European Restaurant          Empanada Restaurant  
1                    Drugstore          Dumpling Restaurant  
2                    Drugstore                      Dog Run  
3          Dumpling Restaurant  Eastern European Restaurant  
4             Doner Restaurant                   Donut Shop

Finally, let's visualize the resulting clusters

In [16]:
# create map
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters

### Cluster 1

In [23]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 0, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood   1st Most Common Venue  \
1              Highland Creek, Rouge Hill, Port Union          History Museum   
2                   Guildwood, Morningside, West Hill          Medical Center   
3                                              Woburn             Coffee Shop   
4                                           Cedarbrae      Athletics & Sports   
6         East Birchmount Park, Ionview, Kennedy Park             Bus Station   
7                     Clairlea, Golden Mile, Oakridge                Bus Line   
8     Cliffcrest, Cliffside, Scarborough Village West                   Motel   
9                         Birch Cliff, Cliffside West         College Stadium   
10  Dorset Park, Scarborough Town Centre, Wexford ...       Indian Restaurant   
11                                  Maryvale, Wexford          Breakfast Spot   
12                                          Agincourt                  Lounge   
13             Clarks Corners, Sullivan, Tam OShanter             Pizza Place   
15                                     LAmoreaux West    Fast Food Restaurant   
16                                  Hillcrest Village      Athletics & Sports   
17                       Fairview, Henry Farm, Oriole          Clothing Store   
18                                    Bayview Village     Japanese Restaurant   
19                                   Willowdale South        Ramen Restaurant   
21                                    Willowdale West             Pizza Place   
23                                    Don Mills North          Baseball Field   
24                   Flemingdon Park, Don Mills South        Asian Restaurant   
25    Bathurst Manor, Downsview North, Wilson Heights             Coffee Shop   
26                    Northwood Park, York University  Furniture / Home Store   
28                                     Downsview West           Shopping Mall   
29                                  Downsview Central          Baseball Field   
30                                Downsview Northwest      Athletics & Sports   
31                                   Victoria Village            Intersection   
32                    Woodbine Gardens, Parkview Hill             Pizza Place   
33                                   Woodbine Heights            Skating Rink   
34                                        The Beaches       Health Food Store   
35                                            Leaside     Sporting Goods Shop   
36                                   Thorncliffe Park       Indian Restaurant   
38                       The Danforth West, Riverdale        Greek Restaurant   
39                     The Beaches West, India Bazaar                    Park   
40                                    Studio District                    Café   
42                                   Davisville North       Food & Drink Shop   
43                                 North Toronto West          Clothing Store   
44                                         Davisville            Dessert Shop   
46  Deer Park, Forest Hill SE, Rathnelly, South Hi...             Coffee Shop   
48                        Cabbagetown, St. James Town             Coffee Shop   
49                               Church and Wellesley             Coffee Shop   
50                          Harbourfront, Regent Park             Coffee Shop   
51                           Ryerson, Garden District             Coffee Shop   
52                                     St. James Town             Coffee Shop   
53                                        Berczy Park             Coffee Shop   
54                                 Central Bay Street             Coffee Shop   
55                           Adelaide, King, Richmond             Coffee Shop   
56  Harbourfront East, Toronto Islands, Union Station             Coffee Shop   
57           Design Exchange, Toronto Dominion Centre             Coffee Shop   
58                     Commerce Court, Victoria Hotel             Coffee

### Cluster 2

In [24]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 1, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
20                                    York Mills West                  Park   
22                                          Parkwoods  Fast Food Restaurant   
27                        CFB Toronto, Downsview East                  Park   
37                                       East Toronto                  Park   
41                                      Lawrence Park                  Park   
47                                           Rosedale                  Park   
71                                Caledonia-Fairbanks                  Park   
87      The Kingsway, Montgomery Road, Old Mill North                  Park   
94                                             Weston                  Park   
96  Kingsview Village, Martin Grove Gardens, Richv...                  Park   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
20                  Bank           Yoga Studio     Electronics Store   
22                  Park     Food & Drink Shop           Yoga Studio   
27               Airport              Bus Stop           Yoga Studio   
37     Convenience Store           Yoga Studio     Electronics Store   
41              Bus Line           Swim School           Yoga Studio   
47                 Trail            Playground           Yoga Studio   
71              Pharmacy  Fast Food Restaurant                Market   
87                 River           Yoga Studio   Dumpling Restaurant   
94     Convenience Store           Yoga Studio     Electronics Store   
96     Mobile Phone Shop              Bus Line           Yoga Studio   

          5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
20             Doner Restaurant            Donut Shop             Drugstore   
22  Eastern European Restaurant               Dog Run      Doner Restaurant   
27  Eastern European Restaurant               Dog Run      Doner Restaurant   
37                      Dog Run      Doner Restaurant            Donut Shop   
41          Dumpling Restaurant               Dog Run      Doner Restaurant   
47               Discount Store               Dog Run      Doner Restaurant   
71  Eastern European Restaurant               Dog Run      Doner Restaurant   
87               Discount Store               Dog Run      Doner Restaurant   
94                      Dog Run      Doner Restaurant            Donut Shop   
96          Dumpling Restaurant               Dog Run      Doner Restaurant   

   8th Most Common Venue        9th Most Common Venue  \
20   Dumpling Restaurant  Eastern European Restaurant   
22            Donut Shop                    Drugstore   
27            Donut Shop                    Drugstore   
37             Drugstore          Dumpling Restaurant   
41            Donut Shop                    Drugstore   
47            Donut Shop                    Drugstore   
71            Donut Shop                    Drugstore   
87            Donut Shop                    Drugstore   
94             Drugstore          Dumpling Restaurant   
96            Donut Shop                    Drugstore   

         10th Most Common Venue  
20          Empanada Restaurant  
22          Dumpling Restaurant  
27          Dumpling Restaurant  
37  Eastern European Restaurant  
41  Eastern European Restaurant  
47          Dumpling Restaurant  
71          Dumpling Restaurant  
87  Eastern European Restaurant  
94  Eastern European Restaurant  
96  Eastern European Restaurant

### Cluster 3

In [25]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 2, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue  \
5                                 Scarborough Village                   Spa   
14  Agincourt North, LAmoreaux East, Milliken, Ste...            Playground   
45                        Moore Park, Summerhill East            Playground   

   2nd Most Common Venue        3rd Most Common Venue  \
5             Playground  Eastern European Restaurant   
14                  Park                  Yoga Studio   
45           Yoga Studio  Eastern European Restaurant   

          4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
5                Discount Store               Dog Run      Doner Restaurant   
14  Eastern European Restaurant        Discount Store               Dog Run   
45               Discount Store               Dog Run      Doner Restaurant   

   7th Most Common Venue 8th Most Common Venue 9th Most Common Venue  \
5             Donut Shop             Drugstore   Dumpling Restaurant   
14      Doner Restaurant            Donut Shop             Drugstore   
45            Donut Shop             Drugstore   Dumpling Restaurant   

   10th Most Common Venue  
5       Electronics Store  
14    Dumpling Restaurant  
45      Electronics Store

### Cluster 4

In [26]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 3, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue 2nd Most Common Venue  \
0  Rouge, Malvern  Fast Food Restaurant           Yoga Studio   

  3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
0     Electronics Store               Dog Run      Doner Restaurant   

  6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
0            Donut Shop             Drugstore   Dumpling Restaurant   

         9th Most Common Venue 10th Most Common Venue  
0  Eastern European Restaurant    Empanada Restaurant

### Cluster 5

In [27]:
toronto_merged.loc[toronto_merged['Cluster Labels'] == 4, toronto_merged.columns[[2] + list(range(6, toronto_merged.shape[1]))]]

Neighborhood 1st Most Common Venue        2nd Most Common Venue  \
60     Roselawn                Garden  Eastern European Restaurant   

   3rd Most Common Venue 4th Most Common Venue 5th Most Common Venue  \
60        Discount Store               Dog Run      Doner Restaurant   

   6th Most Common Venue 7th Most Common Venue 8th Most Common Venue  \
60            Donut Shop             Drugstore   Dumpling Restaurant   

   9th Most Common Venue 10th Most Common Venue  
60           Yoga Studio                  Diner

## Thank you for reviewing my assignment!